File: src/utils/guidanceEngine.js
// ═══════════════════════════════════════════
// Guidance Engine — Post-Step Recommendations
// ═══════════════════════════════════════════

function getTodayStr() {
  return new Date().toISOString().split("T")[0];
}

function daysUntil(dateStr) {
  if (!dateStr) return Infinity;
  const today = new Date(getTodayStr());
  const target = new Date(dateStr);
  return Math.round((target - today) / (1000 * 60 * 60 * 24));
}

/**
 * After completing a step, generate ranked recommendations for what to do next.
 *
 * @param {object} completedStep - The step just completed { id, text, done }
 * @param {object} quest - The quest containing the completed step
 * @param {object[]} allQuests - All quests
 * @param {object} blossomRecs - { readyToAdvance, needsFate, suggestedSeeds } from useBlossomMode
 * @returns {{ recommendations: object[], todayProgress: object, allClear: boolean }}
 */
export function getNextRecommendations(completedStep, quest, allQuests, blossomRecs) {
  const recommendations = [];
  const today = getTodayStr();

  // ── 1. Same quest next step (highest priority — maintain flow) ──
  const nextInQuest = quest.steps.find((s) => !s.done && s.id !== completedStep.id);
  if (nextInQuest) {
    recommendations.push({
      type: "same-quest-step",
      priority: 100,
      questId: quest.id,
      questName: quest.name,
      stepId: nextInQuest.id,
      stepText: nextInQuest.text,
      label: "continue",
    });
  }

  // ── 2. Overdue steps across other quests (urgent) ──
  for (const q of allQuests) {
    if (q.id === quest.id) continue;
    for (const s of q.steps) {
      if (s.done) continue;
      const stepDays = daysUntil(s.deadline);
      if (stepDays < 0) {
        recommendations.push({
          type: "overdue-step",
          priority: 90,
          questId: q.id,
          questName: q.name,
          stepId: s.id,
          stepText: s.text,
          daysOverdue: Math.abs(stepDays),
          label: "overdue",
        });
      }
    }
    // Quest-level overdue
    if (q.deadline && daysUntil(q.deadline) < 0 && !q.steps.every((s) => s.done)) {
      const nextStep = q.steps.find((s) => !s.done);
      if (nextStep && !recommendations.some((r) => r.stepId === nextStep.id)) {
        recommendations.push({
          type: "overdue-quest",
          priority: 88,
          questId: q.id,
          questName: q.name,
          stepId: nextStep.id,
          stepText: nextStep.text,
          daysOverdue: Math.abs(daysUntil(q.deadline)),
          label: "overdue",
        });
      }
    }
  }

  // ── 3. Due-today steps across other quests ──
  for (const q of allQuests) {
    if (q.id === quest.id) continue;
    for (const s of q.steps) {
      if (s.done) continue;
      if (s.deadline === today) {
        recommendations.push({
          type: "today-step",
          priority: 70,
          questId: q.id,
          questName: q.name,
          stepId: s.id,
          stepText: s.text,
          label: "dueToday",
        });
      }
    }
  }

  // ── 4. Blossom recommendations ──
  if (blossomRecs) {
    // Fate decisions needed
    for (const node of (blossomRecs.needsFate || []).slice(0, 2)) {
      recommendations.push({
        type: "blossom-fate",
        priority: 60,
        nodeId: node.id,
        nodeName: node.name,
        label: "blossomFate",
      });
    }
    // Ready to advance
    for (const node of (blossomRecs.readyToAdvance || []).slice(0, 2)) {
      recommendations.push({
        type: "blossom-advance",
        priority: 55,
        nodeId: node.id,
        nodeName: node.name,
        currentStage: node.currentStage,
        nextStage: node.nextStage,
        label: "blossomAdvance",
      });
    }
  }

  // ── 5. Other quests' next steps (lowest priority) ──
  for (const q of allQuests) {
    if (q.id === quest.id) continue;
    if (q.steps.every((s) => s.done)) continue;
    const nextStep = q.steps.find((s) => !s.done);
    if (nextStep && !recommendations.some((r) => r.stepId === nextStep.id)) {
      recommendations.push({
        type: "other-quest-step",
        priority: 30,
        questId: q.id,
        questName: q.name,
        stepId: nextStep.id,
        stepText: nextStep.text,
        label: "otherQuest",
      });
    }
  }

  // Sort by priority descending, take top 4
  recommendations.sort((a, b) => b.priority - a.priority);
  const top = recommendations.slice(0, 4);

  // ── Today's progress ──
  const allSteps = allQuests.flatMap((q) => q.steps);
  const todayDone = allSteps.filter((s) => s.done).length;
  const todayTotal = allSteps.length;
  const todayProgress = { done: todayDone, total: todayTotal };

  // All clear?
  const allClear = allSteps.length > 0 && allSteps.every((s) => s.done);

  return { recommendations: top, todayProgress, allClear };
}

File: ios/QuickTrack/QuickTrack/Engine/GuidanceEngine.swift
import Foundation

/// Generates ranked "what's next" recommendations after a step completes.
/// Ported from web's `src/utils/guidanceEngine.js`.
///
/// Priority order:
///   100: same-quest next step (maintain flow)
///    90: overdue steps in OTHER quests
///    88: overdue quest deadlines
///    70: due-today steps in OTHER quests
///    30: other quests' next steps (fallback)
struct GuidanceEngine {

    enum RecommendationType: String {
        case sameQuestStep      // Highest — continue current quest
        case overdueStep        // Specific overdue step
        case overdueQuest       // Quest with overdue deadline
        case todayStep          // Step due today
        case otherQuestStep     // Generic next step from another quest
    }

    struct Recommendation: Identifiable, Equatable {
        let id: String          // stepId (unique within active steps)
        let type: RecommendationType
        let priority: Int
        let questId: String
        let questName: String
        let stepText: String
        let daysOverdue: Int?   // Only for overdue types

        /// SF Symbol icon based on recommendation type
        var icon: String {
            switch type {
            case .sameQuestStep: "play.fill"
            case .overdueStep, .overdueQuest: "exclamationmark.triangle.fill"
            case .todayStep: "clock.fill"
            case .otherQuestStep: "arrow.right.circle.fill"
            }
        }

        /// Short label describing the urgency
        var urgencyLabel: String {
            switch type {
            case .sameQuestStep:
                return "continue"
            case .overdueStep:
                if let d = daysOverdue {
                    return d == 1 ? "1 day overdue" : "\(d) days overdue"
                }
                return "overdue"
            case .overdueQuest:
                if let d = daysOverdue {
                    return "quest \(d)d overdue"
                }
                return "overdue"
            case .todayStep:
                return "due today"
            case .otherQuestStep:
                return "next up"
            }
        }
    }

    struct GuidanceResult {
        let recommendations: [Recommendation]
        let todayDone: Int
        let todayTotal: Int
        let allClear: Bool
    }

    // MARK: - Main entry

    static func getNextRecommendations(
        completedStep: QuestStep,
        quest: QuestRow,
        allQuests: [QuestRow]
    ) -> GuidanceResult {
        var recs: [Recommendation] = []
        let today = Config.todayString()

        // 1. Same quest next step (highest priority)
        if let nextInQuest = quest.steps.first(where: { !$0.done && $0.id != completedStep.id }) {
            recs.append(Recommendation(
                id: nextInQuest.id,
                type: .sameQuestStep,
                priority: 100,
                questId: quest.id,
                questName: quest.name,
                stepText: nextInQuest.text,
                daysOverdue: nil
            ))
        }

        // 2. Overdue steps in OTHER quests
        for q in allQuests where q.id != quest.id {
            for s in q.steps where !s.done {
                if let stepDeadline = stepDeadline(s),
                   let days = daysBetween(today: today, target: stepDeadline),
                   days < 0 {
                    recs.append(Recommendation(
                        id: s.id,
                        type: .overdueStep,
                        priority: 90,
                        questId: q.id,
                        questName: q.name,
                        stepText: s.text,
                        daysOverdue: abs(days)
                    ))
                }
            }
            // Quest-level overdue
            if let deadline = q.deadline,
               let days = daysBetween(today: today, target: deadline),
               days < 0,
               !q.steps.allSatisfy(\.done),
               let nextStep = q.steps.first(where: { !$0.done }),
               !recs.contains(where: { $0.id == nextStep.id }) {
                recs.append(Recommendation(
                    id: nextStep.id,
                    type: .overdueQuest,
                    priority: 88,
                    questId: q.id,
                    questName: q.name,
                    stepText: nextStep.text,
                    daysOverdue: abs(days)
                ))
            }
        }

        // 3. Due-today steps in OTHER quests
        for q in allQuests where q.id != quest.id {
            for s in q.steps where !s.done {
                if stepDeadline(s) == today,
                   !recs.contains(where: { $0.id == s.id }) {
                    recs.append(Recommendation(
                        id: s.id,
                        type: .todayStep,
                        priority: 70,
                        questId: q.id,
                        questName: q.name,
                        stepText: s.text,
                        daysOverdue: nil
                    ))
                }
            }
        }

        // 4. Other quests' next steps (fallback)
        for q in allQuests where q.id != quest.id {
            if q.steps.allSatisfy(\.done) { continue }
            if let nextStep = q.steps.first(where: { !$0.done }),
               !recs.contains(where: { $0.id == nextStep.id }) {
                recs.append(Recommendation(
                    id: nextStep.id,
                    type: .otherQuestStep,
                    priority: 30,
                    questId: q.id,
                    questName: q.name,
                    stepText: nextStep.text,
                    daysOverdue: nil
                ))
            }
        }

        // Sort by priority desc, take top 3
        recs.sort { $0.priority > $1.priority }
        let top = Array(recs.prefix(3))

        // Today's overall progress
        let allSteps = allQuests.flatMap(\.steps)
        let todayDone = allSteps.filter(\.done).count
        let todayTotal = allSteps.count
        let allClear = !allSteps.isEmpty && allSteps.allSatisfy(\.done)

        return GuidanceResult(
            recommendations: top,
            todayDone: todayDone,
            todayTotal: todayTotal,
            allClear: allClear
        )
    }

    // MARK: - Helpers

    /// QuestStep currently has no `deadline` field in Swift model. If/when added, expose it here.
    /// For now, steps inherit their quest's deadline.
    private static func stepDeadline(_ step: QuestStep) -> String? {
        return nil  // Step-level deadlines not yet modeled in Swift (web has them)
    }

    /// Calendar-day difference between today (YYYY-MM-DD) and target (YYYY-MM-DD).
    /// Positive = target is in the future. Negative = overdue.
    private static func daysBetween(today: String, target: String) -> Int? {
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd"
        fmt.locale = Locale(identifier: "en_US_POSIX")
        guard let todayDate = fmt.date(from: today),
              let targetDate = fmt.date(from: target) else { return nil }
        return Calendar.current.dateComponents([.day], from: todayDate, to: targetDate).day
    }
}
